go/vt/vtgateproxy/discovery.go

```diff
@@ -144,39 +195,55 @@ func RegisterJSONGateResolver(
 	return jsonDiscovery, nil
 }
 
-func (*JSONGateResolverBuilder) Scheme() string { return "vtgate" }
+func (*GateResolverBuilder) Scheme() string { return "vtgate" }
 
-// Parse and validate the format of the file and start watching for changes
-func (b *JSONGateResolverBuilder) start() error {
-	// Perform the initial parse
-	_, err := b.parse()
-	if err != nil {
-		return err
-	}
+func (b *GateResolverBuilder) targetsChanged(allTargets map[string][]targetHost) {
+	b.selectTargets(allTargets)
 
-	// Validate some stats
-	if len(b.targets) == 0 {
-		return fmt.Errorf("no valid targets in file %s", b.jsonPath)
-	}
-
-	// Log some stats on startup
-	poolTypes := map[string]int{}
-	affinityTypes := map[string]int{}
+	var wg sync.WaitGroup
 
-	for _, ts := range b.targets {
-		for _, t := range ts {
-			count := poolTypes[t.PoolType]
-			poolTypes[t.PoolType] = count + 1
+	// notify all the resolvers that the targets changed in parallel, since each update might sleep for
+	// the warmup time
+	b.mu.RLock()
+	for _, r := range b.resolvers {
+		wg.Add(1)
+		go func(r *GateResolver) {
+			defer wg.Done()
 
-			count = affinityTypes[t.Affinity]
-			affinityTypes[t.Affinity] = count + 1
-		}
+			err := b.update(r)
+			if err != nil {
+				log.Errorf("Failed to update resolver: %v", err)
+			}
+		}(r)
 	}
+	b.mu.RUnlock()
+	wg.Wait()
+}
 
-	parseCount.Add("changed", 1)
+func (b *GateResolverBuilder) discoveryLoopRotor() error {
+	// Start a config watcher
+	b.ticker = time.NewTicker(1 * time.Second)
 
-	log.Infof("loaded targets, pool types %v, affinity %s, groups %v", poolTypes, *affinityValue, affinityTypes)
+	go func() {
+		for range b.ticker.C {
 
+			allTargets, err := b.discoverFromRotor()
```

Comment:
Does rotor require polling like this?

If so that seems significantly worse than the model we have today since we're going to be busy polling the big payload every second?